python/pyarrow/_dataset.pyx

```diff
@@ -426,6 +469,32 @@ cdef class FileFragment(Fragment):
         return FileFormat.wrap(self.file_fragment.format())
 
 
+cdef class ParquetFileFragment(FileFragment):
+    """A Fragment representing a parquet file."""
+
+    cdef:
+        CParquetFileFragment* parquet_file_fragment
+        int _num_row_groups
+
+    cdef void init(self, const shared_ptr[CFragment]& sp):
+        FileFragment.init(self, sp)
+        self.parquet_file_fragment = <CParquetFileFragment*> sp.get()
+        self._num_row_groups = -1
+
+    @property
+    def row_groups(self):
+        row_groups = set(self.parquet_file_fragment.row_groups())
+        if len(row_groups) != 0:
+            return row_groups
+
+        if self._num_row_groups == -1:
```

Comment:
Should this be taken care at the C++ level? Looks like internal optimisation leaking (vector of size 0).

Reply:
Or, I think it is fine to simply return `None` here, to indicate that all row groups and not a specific set or row groups will be read.

Reply:
Handling this in C++ is certainly possible, [an earlier draft included this behavior](https://github.com/apache/arrow/pull/6670/commits/351512665814e816ef46ae1dda7fe97a378e3af9#diff-b7aed3b5c0adacff778c6930cc6a03e4R417-R427). However this requires I/O on construction or lazily when the field is accessed (as here). I removed it since it seemed better not to hide implicit I/O in C++

Returning `None` is certainly a possibility. My intent here was to allow a ParquetFileFragment to be easily introspected for its row groups even if all are selected to expedite construction of sub fragments with a narrower selection. For example:

```python
def single_row_group_fragments(parquet_fragment):
  for row_group in parquet_fragment.row_groups:
    yield ParquetFileFragment(parquet_fragment.path, set(row_group))
```

(Note: this PR doen't expose such a constructor for ParquetFileFragment)

Reply:
> My intent here was to allow a ParquetFileFragment to be easily introspected for its row groups even if all are selected

There might be alternative ways to provide similar functionality (eg exposing a ``num_row_groups`` property on the fragment, or actually exposing the parquet metadata from which the number of row groups can be accessed. 

I *think* there is value in having this property indicate whether a subset of row groups is manually specified or not.

Reply:
okay, I'll remove this for now